mountpoint-s3/src/fs/error.rs

```diff
@@ -32,25 +34,32 @@ use crate::upload::UploadWriteError;
 #[macro_export]
 macro_rules! err {
     // Base case -- don't use directly
-    ($errno:expr, __source:$source:expr, $message:literal, $($args:tt)*) => {
+    ($errno:expr, __source:$source:expr, $level:expr, $message:literal, $($args:tt)*) => {
         Error {
             errno: $errno,
             message: format!($message, $($args)*),
             source: $source,
+            level: $level,
         }
     };
     // Actual cases
+    ($errno:expr, source:$source:expr, $level:expr, $message:literal, $($args:tt)*) => {
+        err!($errno, __source:Some(::anyhow::Error::new($source)), $level, $message, $($args)*)
+    };
+    ($errno:expr, source:$source:expr, $level:expr, $message:literal,) => {
+        err!($errno, __source:Some(::anyhow::Error::new($source)), $level, $message,)
+    };
     ($errno:expr, source:$source:expr, $message:literal, $($args:tt)*) => {
-        err!($errno, __source:Some(::anyhow::Error::new($source)), $message, $($args)*)
+        err!($errno, __source:Some(::anyhow::Error::new($source)), ::tracing::Level::WARN, $message, $($args)*)
     };
     ($errno:expr, source:$source:expr, $message:literal) => {
-        err!($errno, __source:Some(::anyhow::Error::new($source)), $message,)
+        err!($errno, __source:Some(::anyhow::Error::new($source)), ::tracing::Level::WARN, $message,)
     };
     ($errno:expr, $message:literal, $($args:tt)*) => {
-        err!($errno, __source:None, $message, $($args)*)
+        err!($errno, __source:None, ::tracing::Level::WARN, $message, $($args)*)
     };
     ($errno:expr, $message:literal) => {
-        err!($errno, __source:None, $message,)
+        err!($errno, __source:None, ::tracing::Level::WARN, $message,)
     };
 }
```

Comment:
What's the benefit of this being a macro? Would we be better served with using `Error` directly, and maybe setters for `source` and `level`?

Usage example...

```rust
let mut fs_err = fs::Error::new(libc::ENOENT, "file does not exist"); // may need `format!(...)` if we want args in there
fs_err.source(err);
fs_err.level(tracing::Level::DEBUG);
return Err(fs_err);
```

Reply:
Any thoughts/preference @passaro / @monthonk?

Reply:
Yes. It can work by using setters. I thought of having the macro as it was already there for other cases. And maybe using macro, we can directly define the error at one go with the number of parameters we want to define.
Also, in the above way, we will have to set a default `source` and `level`. Should we do that?

Reply:
Let's hold off on it for now.

Reply:
My 2 cents. If using setters, how do you ensure, for example that both source and level are set (if that is desired)?

If feels that using the macro also allows more flexibility with the types, for example if we need to change them later.